src/hotspot/share/oops/stackChunkOop.cpp

```diff
@@ -441,32 +441,33 @@ void stackChunkOopDesc::fix_thawed_frame(const frame& f, const RegisterMapT* map
 template void stackChunkOopDesc::fix_thawed_frame(const frame& f, const RegisterMap* map);
 template void stackChunkOopDesc::fix_thawed_frame(const frame& f, const SmallRegisterMap* map);
 
-void stackChunkOopDesc::copy_lockstack(oop* dst) {
-  int cnt = lockstack_size();
-
-  if (!(is_gc_mode() || requires_barriers())) {
-    oop* lockstack_start = (oop*)start_address();
-    for (int i = 0; i < cnt; i++) {
-      dst[i] = lockstack_start[i];
-      assert(oopDesc::is_oop(dst[i]), "not an oop");
+void stackChunkOopDesc::transfer_lockstack(oop* dst) {
+  const bool requires_gc_barriers = is_gc_mode() || requires_barriers();
+  const bool requires_uncompress = has_bitmap() && UseCompressedOops;
+  const auto load_and_clear_obj = [&](intptr_t* at) -> oop {
+    if (requires_gc_barriers) {
+      if (requires_uncompress) {
+        oop value = HeapAccess<>::oop_load(reinterpret_cast<narrowOop*>(at));
+        HeapAccess<>::oop_store(reinterpret_cast<narrowOop*>(at), nullptr);
+        return value;
+      } else {
+        oop value = HeapAccess<>::oop_load(reinterpret_cast<oop*>(at));
+        HeapAccess<>::oop_store(reinterpret_cast<oop*>(at), nullptr);
+        return value;
+      }
+    } else {
+      oop value = *reinterpret_cast<oop*>(at);
+      HeapAccess<>::oop_store(reinterpret_cast<oop*>(at), nullptr);
```

Comment:
Using HeapAccess when `!requires_gc_barriers` is wrong. This would crash with ZGC when/if we fix the flags race and changed `relativize_chunk_concurrently` to only be conditioned `requires_barriers() / _barriers` (and allowing the retry_fast_path "medium" path). 
So either use `*reinterpret_cast<oop*>(at) = nullptr;` or do what my initial suggestion with `clear_lockstack` did, just omit the clearing. Before we requires_barriers(), we are allowed to reuse the stackChuncks, so trying to clean them up seems fruitless.

Reply:
Ok, I just omitted clearing the oop.